design-proposals/2022-11-29-metal-port.md

```diff
@@ -0,0 +1,447 @@
+
+Figure: Heap Support
+
+The good way to do memory management is heaps.  Instead of allocating a buffer in random shared memory, you allocate it out of a Metal Heap.  Metal's attitude toward what's in that Heap is "you do you".  A lot of the guardrails go away and you're very much on your own for managing that memory.  But it is so very, very fast.
+
+The reason Metal Heaps are fast goes back to the Figure: Metal Renderer.  All of those lines related to wiring up buffers get much simpler.  Every time you tell Metal about a bit of memory needed to feed the Shader it has to figure out where that memory is, make sure you're allowed to access it, do the math for the offsets and make sure you're not outside the bounds.  You might pass in a lot of buffers to a particular shader and there are going to be thousands of the things for a given scene.
+
+If we allocate Buffers out of Heaps, then we just need to pass in the Heap and what amounts to an offset.  Metal does much less work for each shader and is less power hungry for the same frame rate.
+
+The drawback is Metal is no longer doing all that checking for us.  So you want a non-Heap version to fall back on when chasing down memory bugs.
+
+#### Required Changes
+
+The developers will probably implement some Buffer allocation infrastructure in a previous phase for Textures and Drawable Buffers.  That will need to be extended to handle Heaps.
+
+We found the following to be useful in WhirlyGlobe-Maply:
+- Allocate a Buffer of a certain size with or without an NSData object to be copied in
+- Allocate a Texture Buffer ofa certain size with or without an NSData object to be copied in
+- Return a handle to describe what was allocated.  If there's additional information, like a Heap, this is where to stash that.
+- If a given object is too large for the pre-allocated Heaps, just create it on its own
+- For a Drawable, reference the Heap and pass in the buffer offset as needed
+
+In the end, heaps are heaps and these are fairly standard
+
+#### Benefits
+
+Managing Metal memory with Heaps is more efficient than individual buffers.  Buffer management tends to dominate the low level wiring of Drawables.  For each frame, much of what you're doing on the CPU is pointing a drawing command at its buffers.  With Heaps, you still do that, but much of the checking goes away.  You make the appropriate Heaps available and then the driver is largely just dealing with offsets.
+
+It's fast and it's a key step to getting as much work as possible off the main CPU.
+
+### Indirect Implementation
+
+The implementation up to this point is Direct.  For each frame we fill out the same command buffers.  Sure, the data buffers are reused, but the command buffers are created, filled in, and handed over to Metal each and every frame.  This is not dissimilar to OpenGL, in a way, but still more efficient.
```

Comment:
My general impression is less fussy setup, more thread support, and a more carefully optimized driver implementation.  

We certainly saw that in our implementation for WhirlyGlobe-Maply.

Reply:
A lot of the overhead from GL is because of the global mutable state, the driver behind the scenes has to do a lot of heavy lifting translating that user flexibility into what the GPU actually wants, _in a way that is also performant_. In modern desktop GL you can work around a lot of the overhead with direct state access and bindless resources.